### DISCUSSION TITLE
How to handle duplicate URLs?

### DISCUSSION THREAD

**Discussion by mrgiao** [plausible/analytics#1544] [Anything Else]
### Problem
Duplicate URLs split statistics. Two types of duplicate URLs - _trailing slash_ and _case sensitive_ - create issues such as:
- Making it harder to get a correct overview of the data.
- More complex when making strategic decision.
- Time consuming when having to aggregate data.
- Confuses stakeholders who are not aware of reasons for different entries.

The owner of a website seldom has control over external link URLs pointing to their website. It is common for links to vary in the correct casing and for links to vary in including/excluding a trailing slash. For example these examples will all generate their own entry for statistics:
- `/about`
- `/about/`
- `/About`
- `/About/`

![image](https://user-images.githubusercontent.com/4223897/146378912-582f7717-a96d-4721-a007-950d1ece9eaf.png)

Even though URLs supposedly are case sensitive and a URL with trailing slash supposedly is unique from the same URL without, they most probably all point to the same page.

### Suggested solution
- Provide a way to merge duplicate URLs.
- Provide a way to unmerge URLs.
- Visually indicate in the statistics that the entry has been merged.
- Provide suggestions for merging similar URLs based on differing cases or trailing slash.

![image](https://user-images.githubusercontent.com/4223897/146381364-eb36b7e0-ca01-4fe8-8253-c083a428e90c.png)

### Further reading
- https://www.getelevar.com/how-to/fix-duplicate-url-google-analytics/
- https://www.searchenginejournal.com/google-urls-are-case-sensitive/421264

**Comment by metmarkosaric** [comment]
Thanks for the summary! 

These are all technical issues with a website that should be fixed on the site itself. Web analytics tools cannot fix these issues.

We may take some action to hide the issues in the future but the issues still remain and should be fixed if you want to have an SEO optimized site. 

Companies pay good money to get technical audits done that discover these issues so in majority of cases it would be wrong of us to try and hide existing issues from people.

There's more on this in https://github.com/plausible/analytics/discussions/336

**Comment by ukutaht** [comment]
The mockup looks really good!

As @metmarkosaric said we'll probably keep showing these all as separate pages by default. That's how URLs are viewed by most tools including search engines. But it would definitely be nice to be able to merge duplicate URLs for reporting purposes.

**Comment by mrgiao** [comment]
Sure I understand why this feature should not be switched on by default if it's the industry standard. But by providing such a feature it can surely only help as it can also expose where this type of problem occurs. Regardless of whether the user is a small business owner or an SEO consultant helping them interpret the data and visitors behavior - including understanding problem areas - can only be a good thing.

![image](https://user-images.githubusercontent.com/4223897/146448635-9112f757-8843-4ba1-a2cc-150ed181b7d5.png)

**Reply by metmarkosaric** [reply]
perhaps we can look at it in the future but there is a whole industry of software tools that specialize in technical issues and it's not normally something that's part of web analytics

**Comment by sahandnayebaziz** [comment]
I just wanted to chime in here to say that I agree with @mrgiao. It would just be really easy for me as a customer if I could fix this problem I currently have with my `/blog` pages by telling Plausible that these two are really supposed to be the same page. 

I get that you are saying that's not really pure or doesn't reflect what other SEO companies do, but that doesn't change that as a customer and fan of Plausible I'm currently scratching my head every time I look at my dashboard wondering what I'm supposed to fix to make those two `/blog` endpoints just count together. (Is it a Netlify problem? Is it a Gatsby problem? Is it a redirects problem?) I'm just sharing that as a user, when I get to Plausible, it's easy for me to consider it a Plausible problem, or wish there was a Plausible solution, because searching for the source of the problem and the solution across my web tech stack ends up not helping out enough times in a row that I stop searching. 

If tomorrow a competitor offered a way to just have the _view_ of that data merge them, that would be one great reason to switch over.  (I like @mrgiao's mockup where you see the URLs being counted together represented in a count at the end of the line)

It's also funny to me a little bit that we are discussing this as a technical problem. Every time I come back to this, I see this as a peculiarity of the internet, search engines, etc and it just seems like a rough edge that I wish was smoothed over. You're probably totally right that it's a technical issue and I'm missing a redirect rule, better canonicals (which I have set up with [gatsby-plugin-canonical-urls](https://www.gatsbyjs.com/plugins/gatsby-plugin-canonical-urls/), or something else, but I still thought it to be valuable to share with you as a user that every few months, I google this, check the Plausible docs, and end up basically leaving it because nothing pops out as helpful or a fix. By definition, a problem I have that someone could sell me a solution for.

<img width="559" alt="Screen Shot 2022-02-06 at 8 58 20 AM" src="https://user-images.githubusercontent.com/909340/152691892-11f6848d-cdfc-4068-9b71-9851b42e83e3.png">

**Reply by metmarkosaric** [reply]
thanks for your feedback @sahandnayebaziz! 

analytics tools just show what happens on your site. if you go on your site with a trailing slash or without one, you get to see two different pages which is why we also show them as they are there. sometimes these two different urls could even have different content. it happens with http/https, www/no-www, trailing slash/no trailing slash, query parameters/no query parameters... they're all part of the same technical issue if the site has them. 

in the future, we may decide to go away from the standards and hide these issues from you, but at the moment we display things as they happen on your site. if we do hide these duplicate pages, we'll make sure to have a note in our reports as most people would not want these types of technical issues being hidden. the best option to fix this problem is to fix it on the site in case you want to optimize your site for things such as search engine rankings. then even your analytics tools we no longer show duplicate pages as they would not exist anymore.

best places for this type of info are seo tools such as ahrefs: https://ahrefs.com/blog/301-redirects/. here's a bit more info from google on this as well that ahrefs linked to: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls. if you're using a cms that doesn't do this out of the box, you can search for 301 redirects for your tool of choice.

**Reply by sahandnayebaziz** [reply]
thank you @metmarkosaric 😊 I really appreciate you taking the time reply, and giving me more context. Thanks for hearing me out on how it "felt" looking it as a customer. Since this morning, I've checked out `ahrefs` which is _super impressive_ and you've got me started on improving the SEO of my site altogether. Really appreciate it.

**Reply by sahandnayebaziz** [reply]
I'm going to head on that path of fixing the duplicates so they don't even show up anymore :)

**Reply by metmarkosaric** [reply]
glad you found that useful @sahandnayebaziz, thanks!

**Reply by hughbris** [reply]
> I'm going to head on that path of fixing the duplicates so they don't even show up anymore :)

That's really the only way and FYI it's not just an SEO issue, it's better for users. Caching and visited link styles benefit users, for example. So I'm much more supportive of leading site developers towards good practices than accommodating poor ones (google actually specialise in that!). If an analytics service were to accommodate duplicate URLs, I'd start by reading the sitemap or canonical URLs from the document's head anyway, rather than building a front end URL matching interface.
